[branch_9_9] Reflow computeCommonPrefixLengthAndBuildHistogram to avoid crash

ChrisHegarty opened the pull request:
This commit reflows the code in the method body of `computeCommonPrefixLengthAndBuildHistogram`, so as to avoid a JVM JIT crash. The purpose of this change is to workaround the JVM bug which is somewhat fragile, but the best that we can do for now and appears to be working well. Further details of the JDK issue can be found in https://bugs.openjdk.org/browse/JDK-8321370 and linked issues.

The newly added test fails around 9 out of 10 times for me, before the fix. With the fix the test has not been seen to fail in several hundreds of runs. Though, further test runs are on going. 


The JDK issue is already fixed in the currently unreleased JDK 21.0.2 and JDK 22, but we need this workaround to avoid the crash on current released JDKs.

closes #12898

ChrisHegarty committed:
Add test that tickles a JVM JIT crash on JDK's less than 21.0.1

ChrisHegarty committed:
reflow computeCommonPrefixLengthAndBuildHistogram

ChrisHegarty commented:
For reviewers or other interested parties. I run the test as follows:

```
export CI=true; export x=; while ./gradlew --no-build-cache cleanTest :lucene:core:test --rerun --tests "org.apache.lucene.util.bkd.TestDocIdsWriter.testCrash"; do echo $x | wc -c; export x=x$x; done
```

We need the `CI=true` because, by default in our gradle environment C2 compilation is disabled - and that is where the JIT crash happens.

jpountz commented on lucene/core/src/test/org/apache/lucene/util/bkd/TestDocIdsWriter.java:
```diff
@@ -150,4 +154,18 @@ public Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
+        }
+      }
+    }
+  }
```
How long does this test take? If it's multiple seconds, should it only run nightly?

ChrisHegarty replied:
The test takes around 3.5 seconds on my laptop. I can probably reduce the iterations. Or make the number of iterations depend on whether nightly or not? I’ll take a look.

ChrisHegarty replied:
Done. [16f91f6](https://github.com/apache/lucene/pull/12903/commits/16f91f6c7dc8a05b3c05de888362999aea01d4f2)

jpountz commented on lucene/core/src/java/org/apache/lucene/util/MSBRadixSorter.java:
```diff
@@ -214,6 +214,12 @@ protected int getBucket(int i, int k) {
+        from, to, k, histogram, commonPrefixLength);
+  }
+
+  private int computeInitialCommonPrefixLength(int from, int k) {
```
Can you add a comment here and in other places that you touched, explaining that this works around a C2 compiler bug?

ChrisHegarty replied:
Yes. I’ll add a comment to that effect.

ChrisHegarty replied:
Done. [16f91f6](https://github.com/apache/lucene/pull/12903/commits/16f91f6c7dc8a05b3c05de888362999aea01d4f2)

dweiss commented on lucene/core/src/test/org/apache/lucene/util/bkd/TestDocIdsWriter.java:
```diff
@@ -150,4 +154,18 @@ public Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
   }
+
+  // This simple test tickles a JVM JIT crash on JDK's less than 21.0.1
+  // Needs to be run with C2, so with the environment variable `CI` set
```
If it's triggered by the CI env. variable (and effectively c2 compilation) then perhaps we should check for System.getProperty("CI") is assumption-ignore the test if it can't be triggered?

dweiss commented:
> export CI=true; export x=; while ./gradlew --no-build-cache cleanTest :lucene:core:test --rerun --tests "org.apache.lucene.util.bkd.TestDocIdsWriter.testCrash"; do echo $x | wc -c; export x=x$x; done

Just a suggestion that this call to gradle should have a similar outcome and be a tad faster (it does fork multiple test JVMs after one compilation - used to run in parallel too but not anymore - gradle's constraint).
```
gradlew --no-build-cache :lucene:core:beast -Ptests.dups=100 --tests "org.apache.lucene.util.bkd.TestDocIdsWriter.testCrash"
```

More info:
https://github.com/apache/lucene/blob/main/help/tests.txt#L89-L123

ChrisHegarty commented:
> ```
> gradlew --no-build-cache :lucene:core:beast -Ptests.dups=100 --tests "org.apache.lucene.util.bkd.TestDocIdsWriter.testCrash"
> ```

Thanks @dweiss , that's nicer, and makes my life a little easier.

ChrisHegarty committed:
Update test

ChrisHegarty committed:
Add comment about the nature of this change

ChrisHegarty commented:
I've addressed all comments so far.  The use of System::getenv is a little annoying, as it required a test-only permission, but it's minimal and make the test very straightforward to run.

dweiss commented on gradle/testing/randomization/policies/tests.policy:
```diff
@@ -63,6 +63,9 @@ grant {
   permission java.lang.RuntimePermission "writeFileDescriptor";
 
+  // needed to check if C2 (implied by the presence of the CI env) is enabled
+  permission java.lang.RuntimePermission "getenv.CI";
```
Ah, right. Yes, this will be needed - sorry I didn't mention it. I think it's fine since we can't detect c2 in any other way (sun.management.compiler is probably not worth relying on).

dweiss reviewed (approved):
Thank you!

uschindler commented:
> I've addressed all comments so far. The use of System::getenv is a little annoying, as it required a test-only permission, but it's minimal and make the test very straightforward to run.

I made a comment in the main PR (use `Constants#IS_CLIENT_VM`, which was added for vectorization API). The `CI` env var is alos not so useful as Jenkins does not set it.
